### PR TITLE
ENH: generator expressions for builtin `all` and `any`

### LINKED ISSUE #1991: FIX: replace list by generator comprehensions in builtin `all` and `any`
... who ever feels bored.

### PR DESCRIPTION
**Pull request by dengemann**
Closes #1991.

### CODE REVIEW

**Comment by dengemann** [comment]
@Eric89GXL ready for review.

**Comment by Eric89GXL** [comment]
Let me know when tests pass, you have some problems with `len`

**Comment by agramfort** [comment]
travis is not happy

**Comment by dengemann** [comment]
@agramfort looks like a false negative.

**Comment by dengemann** [comment]
Good for review / merge. Travis does not know that he is actually happy.

File: mne/channels/layout.py [hunk 1]
```diff
@@ -370,7 +369,7 @@ def find_layout(info, ch_type=None, exclude='bads'):
                        FIFF.FIFFV_COIL_CTF_OFFDIAG_REF_GRAD)
     has_CTF_grad = (FIFF.FIFFV_COIL_CTF_GRAD in coil_types or
                     (FIFF.FIFFV_MEG_CH in channel_types and
-                     any([k in ctf_other_types for k in coil_types])))
+                     any(k in ctf_other_types for k in coil_types)))
     # hack due to MNE-C bug in IO of CTF
     n_kit_grads = len([ch for ch in chs
                        if ch['coil_type'] == FIFF.FIFFV_COIL_KIT_GRAD])
```
Comment (Eric89GXL) [inline mne/channels/layout.py]: If we're really shooting to use generators, could this just be `sum(ch['coil_type'] = FIFF.FIFFV_COIL_KIT_GRAD for ch in chs)`?

**Comment by Eric89GXL** [comment]
I restarted the failed Travis run, LGTM. If you don't feel like changing the `len`s to use `sum` it's fine by me

**Comment by dengemann** [comment]
> . If you don't feel like changing the lens to use sum it's fine by me

no let's do that, it's more elegant

**Comment by Eric89GXL** [comment]
Travis is worthless, but Appveyor has feedback:

https://ci.appveyor.com/project/Eric89GXL/mne-python/build/1.0.1416

**Comment by dengemann** [comment]
@Eric89GXL looking more closely the latest commit is not really good. I'm removing it.

**Comment by Eric89GXL** [comment]
So good to go from your end, not going to use the `sum` trick?

**Comment by dengemann** [comment]
> So good to go from your end, not going to use the sum trick?

does not make sense for this use case, we're not summing numbers.

**Comment by dengemann** [comment]
and ```sum(i for i, _ in enumerate(....))``` is not really easier to read.

**Comment by Eric89GXL** [comment]
You're summing logicals, which should take up less memory than doing a len of a list that only gets the members for which the logical is true...

**Comment by Eric89GXL** [comment]
I was thinking of this case:
```
len([x for x in xx if x == 'foo'])
```
can be more efficiently computed as
```
sum(x == 'foo' for x in xx)
```

**Comment by dengemann** [comment]
> You're summing logicals, which should take up less memory than doing a len of a list that only gets the members for which the logical is true...

it was trings, not?

File: mne/channels/layout.py (hunk 1 above)
Comment (Eric89GXL) [inline mne/channels/layout.py]: ...this case, in other words

**Comment by dengemann** [comment]
@Eric89GXL yes I know but that particular line was not producing logicals

File: mne/channels/layout.py (hunk 1 above)
Comment (dengemann) [inline mne/channels/layout.py]: I should have read your comment ... rather than anticipating what's next

**Comment by Eric89GXL** [comment]
This:
```
     n_kit_grads = len([ch for ch in chs
                        if ch['coil_type'] == FIFF.FIFFV_COIL_KIT_GRAD])
```
should be equivalent to this:
```
     n_kit_grads = sum(ch['coil_type'] == FIFF.FIFFV_COIL_KIT_GRAD for ch in chs)
```

**Comment by Eric89GXL** [comment]
I'm actually not sure which would be more efficient in all cases, but the second one at least uses a generator :)

**Comment by Eric89GXL** [comment]
Are you working on another commit? If not, I'll test locally and merge if all is good.

**Comment by dengemann** [comment]
forgot to push

**Comment by Eric89GXL** [comment]
Cool, thanks

**Pull request merged**